lib/spack/spack/cray_manifest.py

```diff
@@ -6,93 +6,105 @@
         )
 
 
-def compiler_from_entry(entry: dict, manifest_path: str):
+def compiler_from_entry(entry: dict, *, manifest_path: str) -> Optional[spack.spec.Spec]:
     # Note that manifest_path is only passed here to compose a
     # useful warning message when paths appear to be missing.
     compiler_name = translated_compiler_name(entry["name"])
-
-    if "prefix" in entry:
-        prefix = entry["prefix"]
-        paths = dict(
-            (lang, os.path.join(prefix, relpath))
-            for (lang, relpath) in entry["executables"].items()
-        )
-    else:
-        paths = entry["executables"]
+    paths = extract_compiler_paths(entry)
 
     # Do a check for missing paths. Note that this isn't possible for
     # all compiler entries, since their "paths" might actually be
     # exe names like "cc" that depend on modules being loaded. Cray
     # manifest entries are always paths though.
-    missing_paths = []
-    for path in paths.values():
-        if not os.path.exists(path):
-            missing_paths.append(path)
-
-    # to instantiate a compiler class we may need a concrete version:
-    version = "={}".format(entry["version"])
-    arch = entry["arch"]
-    operating_system = arch["os"]
-    target = arch["target"]
-
-    compiler_cls = spack.compilers.class_for_compiler_name(compiler_name)
-    spec = spack.spec.CompilerSpec(compiler_cls.name, version)
-    path_list = [paths.get(x, None) for x in ("cc", "cxx", "f77", "fc")]
-
+    missing_paths = [x for x in paths if not os.path.exists(x)]
     if missing_paths:
         warnings.warn(
             "Manifest entry refers to nonexistent paths:\n\t"
             + "\n\t".join(missing_paths)
-            + f"\nfor {str(spec)}"
+            + f"\nfor {entry['name']}@{entry['version']}"
             + f"\nin {manifest_path}"
             + "\nPlease report this issue"
         )
 
-    return compiler_cls(spec, operating_system, target, path_list)
+    try:
+        compiler_spec = compiler_spec_from_paths(pkg_name=compiler_name, compiler_paths=paths)
+    except spack.error.SpackError as e:
+        tty.debug(f"[CRAY MANIFEST] {e}")
+        return None
+
+    compiler_spec.constrain(
+        f"platform=linux os={entry['arch']['os']} target={entry['arch']['target']}"
+    )
+    return compiler_spec
+
+
+def compiler_spec_from_paths(*, pkg_name: str, compiler_paths: Iterable[str]) -> spack.spec.Spec:
+    """Returns the external spec associated with a series of compilers, if any."""
+    pkg_cls = spack.repo.PATH.get_pkg_class(pkg_name)
+    finder = ExecutablesFinder()
+    specs = finder.detect_specs(pkg=pkg_cls, paths=compiler_paths)
+
+    if not specs or len(specs) > 1:
+        raise CrayCompilerDetectionError(
+            message=f"cannot detect a single {pkg_name} compiler for Cray manifest entry",
+            long_message=f"Analyzed paths are: {', '.join(compiler_paths)}",
+        )
+
+    return specs[0]
```

Comment:
I wonder if this stuff could be integrated as more of an extension to `spack external find -t compilers`, and if `spack compiler find` could be a pure alias for that command.  I think the behavior is different right now if you go through `spack external find` vs. `spack compiler find` right?

Reply:
> I think the behavior is different right now if you go through spack external find vs. spack compiler find right?

Yes, there are a few slight differences on default search paths, which I maintained from `develop`